Don't absorb constant data unless explicitly tagged

mcr229 opened the pull request:
Summary:
As long as constant data is used as input to a partitioned graph, it will be absorbed by the partitioned edge program. Meaning all constant data would be moved from the owning program to the partitioned program. 

Delegates owning data is not always a desired affect. Delegates should instead be able to choose whether or not they want to take ownership of the data by explicitly tagging the constant data nodes. If they are tagged, then transfer ownership to the delegate, if they are not tagged, then constant data will be owned by overarching edge_program and fed in as inputs to the delegate.

This changes enables this functionality. Note that delegates owning data now must be explicitly tagged. Delegates with partitioners which previously did not tag data will now have to do so in order to preserve their functionality.

Differential Revision: D54874255

YifanShenSZ commented on backends/apple/coreml/partition/coreml_partitioner.py:
```diff
@@ -80,6 +81,8 @@ def partition(self, exported_program: ExportedProgram) -> PartitionResult:
                 node.meta["delegation_tag"] = tag
                 partition_tags[tag] = self.delegation_spec
 
+        tag_constant_data(exported_program)
```
Trying to understandt:
1. So for now we just tag all constant data as supported by backend, right?
2. If we want to selectively tag consts, we will use something like `is_param(edge_exported_program, node) or is_buffer(edge_exported_program, node)` then tag ourselves?

mcr229 replied:
1. Yea, to maintain functionality, we tag_constant_data tags all the constant data which is used by partitions. This maintains the functionality that we currently have.
2. Yea if you want to selectively tag consts you can filter it however you like, whether using is_param or is_buffer, and add the tag. The call `tag_constant_data`, tags all the constant data which are used by the partitions. This maintains the functionality as before. If you don't want to handle owning and serializing the data, you can choose to not tag the data node, and instead the data node will be fed as an input to the delegate.

Jerry-Ge commented on backends/arm/arm_partitioner.py:
```diff
@@ -86,6 +87,8 @@ def partition(self, exported_program: ExportedProgram) -> PartitionResult:
                 node.meta["delegation_tag"] = tag
                 partition_tags[tag] = self.delegation_spec
 
+        tag_constant_data(exported_program)
```
this change isn't changing arm backend's functionality, looks good to me.

chunit-quic commented on backends/qualcomm/partition/qnn_partitioner.py:
```diff
@@ -137,6 +138,7 @@ def partition(self, edge_program: torch.export.ExportedProgram) -> PartitionResu
         partitions = self.generate_partitions(edge_program)
         if len(partitions) != 0:
             self.tag_nodes(partitions)
+            tag_constant_data(edge_program)
```
It seems that no changing to us. Will also test locally. LGTE.

robell reviewed (approved):
LGTM.

mcr229 committed:
Don't absorb constant data unless explicitly tagged (#2424)

Summary:

As long as constant data is used as input to a partitioned graph, it will be absorbed by the partitioned edge program. Meaning all constant data would be moved from the owning program to the partitioned program. 

Delegates owning data is not always a desired affect. Delegates should instead be able to choose whether or not they want to take ownership of the data by explicitly tagging the constant data nodes. If they are tagged, then transfer ownership to the delegate, if they are not tagged, then constant data will be owned by overarching edge_program and fed in as inputs to the delegate.

This changes enables this functionality. Note that delegates owning data now must be explicitly tagged. Delegates with partitioners which previously did not tag data will now have to do so in order to preserve their functionality.

Reviewed By: cccclai

Differential Revision: D54874255

cccclai commented:
https://github.com/pytorch/executorch/pull/2570 is the follow up to address the needs to duplicate constant nodes if needed
